File: hw/toapanta/u3/hw35CrudStrategy/CrudStrategyPY/model/strategy/json_strategy.py
import json
import os
from typing import List
from model.customer import Customer
from model.strategy.crud_strategy import CrudStrategy
from utils.file_manager import FileManager

class JsonStrategy(CrudStrategy):
    FILE_PATH = "customers.json"
    
    def add(self, customer: Customer) -> bool:
        try:
            customers = self.read_all()
            
            for c in customers:
                if c.id == customer.id:
                    print(f"Error: ID {customer.id} already exists in JSON")
                    return False
            
            customers.append(customer)
            
            customers_dict = [c.to_dict() for c in customers]
            json_data = json.dumps(customers_dict, indent=4)
            
            FileManager.save_to_file(self.FILE_PATH, json_data)
            return True
            
        except Exception as e:
            print(f"Error adding customer to JSON: {e}")
            return False
    
    def delete(self, customer_id: int) -> bool:
        try:
            customers = self.read_all()
            initial_count = len(customers)
            
            customers = [c for c in customers if c.id != customer_id]
            
            if len(customers) < initial_count:
                customers_dict = [c.to_dict() for c in customers]
                json_data = json.dumps(customers_dict, indent=4)
                FileManager.save_to_file(self.FILE_PATH, json_data)
                return True
            
            return False
            
        except Exception as e:
            print(f"Error deleting from JSON: {e}")
            return False
    
    def update(self, customer_id: int, customer: Customer) -> bool:
        try:
            customers = self.read_all()
            updated = False
            
            for i, c in enumerate(customers):
                if c.id == customer_id:
                    customer.id = customer_id  
                    customers[i] = customer
                    updated = True
                    break
            
            if updated:
                customers_dict = [c.to_dict() for c in customers]
                json_data = json.dumps(customers_dict, indent=4)
                FileManager.save_to_file(self.FILE_PATH, json_data)
            
            return updated
            
        except Exception as e:
            print(f"Error updating in JSON: {e}")
            return False
    
    def read_all(self) -> List[Customer]:
        try:
            content = FileManager.read_file(self.FILE_PATH)
            if not content:
                return []
            
            customers_dict = json.loads(content)
            customers = [Customer.from_dict(c) for c in customers_dict]
            return customers
            
        except Exception as e:
            print(f"Error reading JSON: {e}")
            return []
    
    def read_by_id(self, customer_id: int) -> Customer:
        customers = self.read_all()
        for customer in customers:
            if customer.id == customer_id:
                return customer
        return None
    
    def get_format_name(self) -> str:
        return "JSON"

File: hw/pastor/u3/hw35CrudStrategy/CrudStrategy/src/main/java/ec/edu/espe/crudstrategy/model/JsonStrategy.java
package ec.edu.espe.crudstrategy.model;

/**
 *
 * @author Mathews Pastor, The POOwer Rangers Of Programming
 */
import ec.edu.espe.crudstrategy.model.Customer;
import ec.edu.espe.crudstrategy.utils.FileManager;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONObject;

public class JsonStrategy implements CrudStrategy {

    private static final String FILE_PATH = "customers.json";

    @Override
    public boolean add(Customer customer) {
        try {
            List<Customer> customers = readAll();
            customers.add(customer);

            JSONArray jsonArray = new JSONArray();
            for (Customer c : customers) {
                JSONObject obj = new JSONObject();
                obj.put("id", c.getId());
                obj.put("name", c.getName());
                obj.put("email", c.getEmail());
                obj.put("phone", c.getPhone());
                jsonArray.put(obj);
            }

            FileManager.saveToFile(FILE_PATH, jsonArray.toString(4));
            return true;
        } catch (Exception e) {
            System.err.println("Error adding customer to JSON: " + e.getMessage());
            return false;
        }
    }

    @Override
    public boolean delete(int id) {
        try {
            List<Customer> customers = readAll();
            boolean removed = customers.removeIf(c -> c.getId() == id);

            if (removed) {
                JSONArray jsonArray = new JSONArray();
                for (Customer c : customers) {
                    JSONObject obj = new JSONObject();
                    obj.put("id", c.getId());
                    obj.put("name", c.getName());
                    obj.put("email", c.getEmail());
                    obj.put("phone", c.getPhone());
                    jsonArray.put(obj);
                }

                FileManager.saveToFile(FILE_PATH, jsonArray.toString(4));
            }
            return removed;
        } catch (Exception e) {
            System.err.println("Error deleting from JSON: " + e.getMessage());
            return false;
        }
    }

    @Override
    public boolean update(int id, Customer updatedCustomer) {
        try {
            List<Customer> customers = readAll();
            boolean updated = false;

            for (int i = 0; i < customers.size(); i++) {
                if (customers.get(i).getId() == id) {
                    customers.set(i, updatedCustomer);
                    updated = true;
                    break;
                }
            }

            if (updated) {
                JSONArray jsonArray = new JSONArray();
                for (Customer c : customers) {
                    JSONObject obj = new JSONObject();
                    obj.put("id", c.getId());
                    obj.put("name", c.getName());
                    obj.put("email", c.getEmail());
                    obj.put("phone", c.getPhone());
                    jsonArray.put(obj);
                }

                FileManager.saveToFile(FILE_PATH, jsonArray.toString(4));
            }
            return updated;
        } catch (Exception e) {
            System.err.println("Error updating in JSON: " + e.getMessage());
            return false;
        }
    }

    @Override
    public List<Customer> readAll() {
        List<Customer> customers = new ArrayList<>();
        try {
            String content = FileManager.readFile(FILE_PATH);
            if (content != null && !content.trim().isEmpty()) {
                JSONArray jsonArray = new JSONArray(content);
                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject obj = jsonArray.getJSONObject(i);
                    Customer customer = new Customer(
                            obj.getInt("id"),
                            obj.getString("name"),
                            obj.getString("email"),
                            obj.getString("phone")
                    );
                    customers.add(customer);
                }
            }
        } catch (Exception e) {
            System.err.println("Error reading JSON: " + e.getMessage());
        }
        return customers;
    }

    @Override
    public Customer readById(int id) {
        return readAll().stream()
                .filter(c -> c.getId() == id)
                .findFirst()
                .orElse(null);
    }

    @Override
    public String getFormatName() {
        return "JSON";
    }
}
